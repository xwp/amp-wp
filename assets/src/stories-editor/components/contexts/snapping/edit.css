.amp-story-page-snap-lines {
	position: absolute;
	top: 0;
	pointer-events: none;

	/* higher than draggable overlay */
	z-index: 20;
}

.amp-story-page-snap-lines line {
	stroke: #f00;
	pointer-events: none;
}
