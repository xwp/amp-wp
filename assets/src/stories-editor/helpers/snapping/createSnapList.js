/**
 * WordPress dependencies
 */
import isShallowEqual from '@wordpress/is-shallow-equal';

/**
 * Returns the setter used for the proxied snap target objects.
 *
 * The setter is a small helper that:
 *
 * - Prevents duplicates
 * - Keeps snap targets within lower and upper bounds.
 * - Combines snap lines if there are multiple for a given target
 *
 * @param {number} lowerBound Lower limit for snap targets, i.e. the page dimensions.
 * @param {number} upperBound Upper limit for snap targets, i.e. the page dimensions.
 * @return {Function} Proxy setter.
 */
const getSetter = ( lowerBound, upperBound ) => ( obj, prop, value ) => {
	prop = Math.round( prop );

	if ( prop < lowerBound || prop > upperBound ) {
		// Discard any snap lines outside the page.
		return true;
	}

	const hasSnapLine = ( item ) =>
		obj[ prop ]
			.find( ( snapLine ) =>
				isShallowEqual( item[ 0 ], snapLine[ 0 ] ) &&
				isShallowEqual( item[ 1 ], snapLine[ 1 ] ),
			);

	if ( obj.hasOwnProperty( prop ) ) {
		if ( hasSnapLine( value ) ) {
			// We already have a completely identical snap line.
			return true;
		}

		// We have at least one snap line at this position, add new to the list.
		obj[ prop ].push( value );
	} else {
		// First snap lines at this position, create new list.
		obj[ prop ] = [ value ];
	}

	// Always keep the list sorted.
	obj[ prop ] = obj[ prop ].sort();

	return true;
};

/**
 * Create snap list via proxy that contains initial snap lines for edges and center of page.
 *
 * @param {import('./types').LineCreator} getLine       Function to create lines based off maximum value.
 * @param {number}                        maxValue      Maximum value of page in this direction.
 * @param {boolean}                       includeEdges  Include edges in initial list.
 * @return {import('./types').SnapLines} An initial list of snap lines ready to be extended.
 */
const createSnapList = ( getLine, maxValue, includeEdges = true ) => new Proxy(
	// Create initial list of snap lines.
	{
		// Center of the page.
		[ maxValue / 2 ]: [ getLine( maxValue / 2 ) ],
		...( includeEdges ?
			{
				// Start page border.
				0: [ getLine( 0 ) ],
				// End page border.
				[ maxValue ]: [ getLine( maxValue ) ],
			} :
			{}
		),
	},
	{
		set: getSetter( 0, maxValue ),
	},
);

export default createSnapList;
