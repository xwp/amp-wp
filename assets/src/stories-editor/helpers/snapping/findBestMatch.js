/**
 * Get the best match for a list of values each creating a match
 *
 * @param {Array.<Object>} targets List of values and matchers to find a best match between.
 * @return {Object} The single best matching object with value and distance (and any extra props from target object)
 */
const findBestMatch = ( ...targets ) => targets
	.map( ( { value, matcher, ...rest } ) => {
		const best = matcher( value );
		if ( ! best ) {
			return null;
		}
		return { value: best, distance: Math.abs( best - value ), ...rest };
	} )
	.filter( Boolean )
	.reduce(
		( best, candidate ) => candidate.distance < best.distance ? candidate : best,
		{ distance: Number.MAX_VALUE, value: null },
	);

export default findBestMatch;
