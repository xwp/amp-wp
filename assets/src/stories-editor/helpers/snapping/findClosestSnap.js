/**
 * External dependencies
 */
import memize from 'memize';

/**
 * Given a number, finds the closest snap target.
 *
 * @see https://github.com/bokuweb/re-resizable
 *
 * @param {number} number Given number.
 * @param {Array.<number> | Function.<number>} snap List of snap targets or function that provides them.
 * @param {number} snapGap Minimum gap required in order to move to the next snapping target
 * @return {?number} Snap target if found.
 */
const findClosestSnap = memize( ( number, snap, snapGap ) => {
	const snapArray = typeof snap === 'function' ? snap( number ) : snap;

	const closestGapIndex = snapArray.reduce(
		( prev, curr, index ) => ( Math.abs( curr - number ) < Math.abs( snapArray[ prev ] - number ) ? index : prev ),
		0,
	);
	const gap = Math.abs( snapArray[ closestGapIndex ] - number );

	return snapGap === 0 || gap < snapGap ? snapArray[ closestGapIndex ] : null;
} );

export default findClosestSnap;
