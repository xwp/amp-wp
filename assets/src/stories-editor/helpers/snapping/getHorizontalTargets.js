/**
 * Internal dependencies
 */
import { STORY_PAGE_INNER_HEIGHT, STORY_PAGE_INNER_WIDTH } from '../../constants';
import getSnapCalculatorByDimension from './getSnapCalculatorByDimension';

const getVerticalLine = (
	offsetX,
	start = 0,
	end = STORY_PAGE_INNER_HEIGHT,
) => [ [ offsetX, start ], [ offsetX, end ] ];

const getHorizontalTargets = getSnapCalculatorByDimension(
	getVerticalLine,
	STORY_PAGE_INNER_WIDTH,
	[ 'left', 'right' ],
	[ 'top', 'bottom' ],
);

export default getHorizontalTargets;
