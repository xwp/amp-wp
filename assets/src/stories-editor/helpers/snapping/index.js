export { default as findClosestSnap } from './findClosestSnap';
export { default as getHorizontalTargets } from './getHorizontalTargets';
export { default as getVerticalTargets } from './getVerticalTargets';
export { default as getBestSnapLines } from './getBestSnapLines';
