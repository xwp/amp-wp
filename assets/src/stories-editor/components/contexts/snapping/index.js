/**
 * External dependencies
 */
import PropTypes from 'prop-types';
import debounce from 'lodash/debounce';

/**
 * WordPress dependencies
 */
import { SVG } from '@wordpress/components';
import { createHigherOrderComponent } from '@wordpress/compose';
import { createContext, useState } from '@wordpress/element';

/**
 * Internal dependencies
 */
import { STORY_PAGE_INNER_HEIGHT, STORY_PAGE_INNER_WIDTH } from '../../../constants';
import './edit.css';

const SnapContext = createContext();

const Snapping = ( { children } ) => {
	const [ snapLines, setRawSnapLines ] = useState( [] );
	const setSnapLines = debounce( setRawSnapLines, 10 );

	const clearSnapLines = () => setSnapLines( [] );

	const context = {
		setSnapLines,
		clearSnapLines,
	};

	return (
		<SnapContext.Provider value={ context }>
			{ children }
			{ Boolean( snapLines.length ) && (
				<SVG
					viewBox={ `0 0 ${ STORY_PAGE_INNER_WIDTH } ${ STORY_PAGE_INNER_HEIGHT }` }
					className="amp-story-page-snap-lines"
				>
					{ snapLines.map( ( [ [ x1, y1 ], [ x2, y2 ] ], index ) => (
						<line
							key={ index }
							x1={ x1 }
							y1={ y1 }
							x2={ x2 }
							y2={ y2 }
						/>
					) ) }
				</SVG>
			) }
		</SnapContext.Provider>
	);
};

Snapping.propTypes = {
	children: PropTypes.object,
};

export default Snapping;

export const withSnapContext = createHigherOrderComponent(
	( WrappedComponent ) => ( props ) => (
		<SnapContext.Consumer>
			{
				( snappingProps ) => (
					<WrappedComponent { ...props } { ...snappingProps } />
				)
			}
		</SnapContext.Consumer>
	),
	'withSnapContext',
);
