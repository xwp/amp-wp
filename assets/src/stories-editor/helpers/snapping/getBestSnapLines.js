/**
 * Internal dependencies
 */
import findClosestSnap from './findClosestSnap';
import findBestMatch from './findBestMatch';

/**
 * Get the best match(es) from a list of snap lines based on actual position.
 *
 * @param {import('./types').SnapLines} edgeLines Object with lists of snap lines indexes by coordinate for edges of object.
 * @param {import('./types').SnapLines} centerLines Object with lists of snap lines indexes by coordinate for center of object.
 * @param {number} start Value of near edge of current object
 * @param {number} end Value of far edge of current object
 * @param {number} gap Maximum gap from edge to snap line for line to be considered.
 * @return {Array.<import('./types').Line>} The lines that correspond to the single best match for any edge.
 */
const getBestSnapLines = ( edgeLines, centerLines, start, end, gap ) => {
	// Go through all snap targets and find the one that is closest.
	const edgeTargets = Object.keys( edgeLines );
	const centerTargets = Object.keys( centerLines );

	const center = ( start + end ) / 2;

	const matcher = ( targets ) => ( value ) => findClosestSnap( value, targets, gap );

	const { value, isEdgeTarget } = findBestMatch(
		{ value: start, matcher: matcher( edgeTargets ), isEdgeTarget: true },
		{ value: end, matcher: matcher( edgeTargets ), isEdgeTarget: true },
		{ value: center, matcher: matcher( centerTargets ), isEdgeTarget: false },
	);

	if ( ! value ) {
		return [];
	}

	return isEdgeTarget ? edgeLines[ value ] : centerLines[ value ];
};

export default getBestSnapLines;
