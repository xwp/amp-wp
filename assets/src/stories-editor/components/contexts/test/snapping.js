/**
 * External dependencies
 */
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';

/**
 * Internal dependencies
 */
import { default as Snapping, withSnapContext } from '../snapping';

// Mock debounce by just returning the original function
jest.mock( 'lodash/debounce', () => ( f ) => f );

const setup = () => {
	const Dummy = () => null;
	// Disable reason: I have no idea why eslint thinks this variable is unused?
	// eslint-disable-next-line @wordpress/no-unused-vars-before-return
	const SnappyDummy = withSnapContext( Dummy );
	const container = (
		<Snapping>
			<SnappyDummy />
		</Snapping>
	);
	const wrapper = mount( container );
	const snapProps = wrapper.find( Dummy ).props();
	const callbacks = [
		'setSnapLines',
		'clearSnapLines',
	];

	// Wrap each callback in act() and follow it up with wrapper.update()
	const wrappedSnapCallbacks = callbacks.reduce(
		( obj, cb ) => ( {
			...obj,
			[ cb ]: ( ...args ) => {
				act( () => snapProps[ cb ]( ...args ) );
				wrapper.update();
			},
		} ),
		{},
	);

	const getDisplayedSnapLines = () => wrapper.find( 'line' );

	return {
		container,
		wrapper,
		getDisplayedSnapLines,
		...wrappedSnapCallbacks,
	};
};

describe( 'Snapping', () => {
	const HORIZONTAL_SNAP_LINE = [ [ 0, 100 ], [ 100, 100 ] ];
	const VERTICAL_SNAP_LINE = [ [ 100, 0 ], [ 100, 100 ] ];

	it( 'should not display any snap lines by default', () => {
		const { getDisplayedSnapLines } = setup();

		const displayedSnapLines = getDisplayedSnapLines();

		expect( displayedSnapLines ).toHaveLength( 0 );
	} );

	it( 'should display snap lines when set', () => {
		const { setSnapLines, getDisplayedSnapLines } = setup();

		setSnapLines( [ VERTICAL_SNAP_LINE, HORIZONTAL_SNAP_LINE ] );

		const displayedSnapLines = getDisplayedSnapLines();

		expect( displayedSnapLines ).toHaveLength( 2 );
	} );

	it( 'should render a single snap line correctly when set', () => {
		const { setSnapLines, getDisplayedSnapLines } = setup();

		setSnapLines( [ HORIZONTAL_SNAP_LINE ] );

		const displayedSnapLines = getDisplayedSnapLines();

		expect( displayedSnapLines ).toMatchSnapshot();
	} );

	it( 'should display only new snap lines when set and set again', () => {
		const { setSnapLines, getDisplayedSnapLines } = setup();

		setSnapLines( [ VERTICAL_SNAP_LINE, HORIZONTAL_SNAP_LINE ] );
		setSnapLines( [ VERTICAL_SNAP_LINE ] );

		const displayedSnapLines = getDisplayedSnapLines();

		expect( displayedSnapLines ).toHaveLength( 1 );
	} );

	it( 'should not display any snap lines when set and cleared', () => {
		const {
			setSnapLines,
			clearSnapLines,
			getDisplayedSnapLines,
		} = setup();

		setSnapLines( [ VERTICAL_SNAP_LINE, HORIZONTAL_SNAP_LINE ] );
		clearSnapLines();

		const displayedSnapLines = getDisplayedSnapLines();

		expect( displayedSnapLines ).toHaveLength( 0 );
	} );
} );
