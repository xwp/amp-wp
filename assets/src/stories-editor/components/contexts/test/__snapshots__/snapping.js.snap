// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapping should render a single snap line correctly when set 1`] = `
<line
  key="0"
  x1={0}
  x2={100}
  y1={100}
  y2={100}
/>
`;
